### PR TITLE
Feature extract method update

### PR DESCRIPTION
**Pull request by bclothier**
This update includes some refactoring, and improved UI logic, notably warning about compilation directives. Also fix the repainting bug when moving the form and enabling resizing of the form.

### CODE REVIEW

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 1]
```diff
@@ -2,69 +2,24 @@
...
-                yield return selector(first, last);
-            }
-        }
-    }
-
     public class ExtractMethodModel
     {
+        private List<string> _fieldsList;
+        private List<string> _parametersList;
+        private List<string> _variablesList;
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: the List suffix is somewhat hungarian-y

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 2]
```diff
@@ -122,11 +78,19 @@ private void Setup()
...
 
             SelectedCode = string.Join(Environment.NewLine, SelectedContexts.Select(c => c.GetText()));
 
+            var rawCode = string.Join(Environment.NewLine,
+                CodeModule.GetLines(1, CodeModule.CountOfLines));
+            var regex = new Regex(@"^(\s?)+#", RegexOptions.Multiline);
+            if (regex.Matches(rawCode).Count > 0)
+            {
+                ModuleContainsCompilationDirectives = true;
+            }
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: can be `ModuleContainsCompilationDirectives = regex.Matches(rawCode).Count > 0` 

Also consider extracting this into an extension method on CodeModules to encapsulate the access of rawCode

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 3]
```diff
@@ -141,19 +105,19 @@ private void Setup()
...
                 {
-                    _parameters = new ObservableCollection<ExtractedParameter>();
+                    _parameters = new ObservableCollection<ExtractMethodParameter>();
                     foreach (var declaration in SourceVariables)
                     {
-                        _parameters.Add(new ExtractedParameter(declaration.AsTypeNameWithoutArrayDesignator,
-                            ExtractParameterNewType.PrivateLocalVariable,
-                            string.Concat(declaration.IdentifierName, declaration.IsArray ? "()" : string.Empty)));
+                        _parameters.Add(new ExtractMethodParameter(declaration.AsTypeNameWithoutArrayDesignator,
+                            ExtractMethodParameterType.PrivateLocalVariable,
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: As mentioned in chat: I'd prefer to pass as parameter by default and allowing users to manually change to local variables. @rubberduck-vba/contributors opinions?

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 4]
```diff
@@ -165,55 +129,62 @@ public string PreviewCode
...
                     {
-                        case ExtractParameterNewType.PublicModuleField:
-                            fields.Add(String.Format($"{Tokens.Public} {parameter.Name} {Tokens.As} {parameter.TypeName}"));
+                        case ExtractMethodParameterType.PublicModuleField:
+                            _fieldsList.Add(parameter.ToString(ExtractMethodParameterFormat.DimOrParameterDeclarationWithAccessibility));
                             break;
-                        case ExtractParameterNewType.PrivateModuleField:
-                            fields.Add(String.Format($"{Tokens.Private} {parameter.Name} {Tokens.As} {parameter.TypeName}"));
+                        case ExtractMethodParameterType.PrivateModuleField:
+                            _fieldsList.Add(parameter.ToString(ExtractMethodParameterFormat.DimOrParameterDeclarationWithAccessibility));
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: consider collapsing case-blocks with the same content:

    case ExtractMethodParameterType.PublicModuleField:
    case ExtractMethodParameterType.PrivateModuleField:
        _fieldsList[....];
        break;

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 5]
```diff
@@ -165,55 +129,62 @@ public string PreviewCode
...
                             break;
                         default:
                             throw new InvalidOperationException("Invalid value for ExtractParameterNewType");
                     }
                 }
 
-                var isFunction = ReturnParameter != null;
+                var isFunction = ReturnParameter != null &&
+                                 !(ReturnParameter.TypeName == string.Empty &&
+                                   ReturnParameter.Name == RubberduckUI.ExtractMethod_NoneSelected);
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: Null-Object pattern anyone?

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs [hunk 6]
```diff
@@ -165,55 +129,62 @@ public string PreviewCode
...
-                strings.AddRange(variables);
-                if (variables.Any())
+                strings.Add(
+                    $@"{Tokens.Private} {(isFunction ? Tokens.Function : Tokens.Sub)} {
+                            NewMethodName ?? RubberduckUI.ExtractMethod_DefaultNewMethodName
+                        }({string.Join(", ", _parametersList)}) {
+                            (isFunction
+                                ? string.Concat(Tokens.As, " ", ReturnParameter.ToString(ExtractMethodParameterFormat.ReturnDeclaration) ?? Tokens.Variant)
+                                : string.Empty)
+                        }");
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodModel.cs]: this one seems somewhat hard to follow. maybe it's better with cleaner syntax-highlighting...

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs [hunk 7]
```diff
@@ -94,13 +94,24 @@ public bool ValidateSelection(QualifiedSelection qualifiedSelection)
...
             {
-                // We've provved that there are no invalid statements contained in the selection. However, we need to analyze
+                // We need to inspect the text directly from the code module because the parse tree do not have compilation
+                // constants or directives easily available (they are in a hidden channel and are encapsulated). Fortunately,
+                // finding compilation constants or directives are easy; they must be prefixed by a "#" and can only have
+                // whitespace between it and the start of the line. Labels or line numbers are not legal in those lines.
+                var rawCode = string.Join(Environment.NewLine,
+                    _codeModule.GetLines(qualifiedSelection.Selection));
+                var regex = new Regex(@"^(\s?)+#", RegexOptions.Multiline);
+                if (regex.Matches(rawCode).Count > 0)
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs]: yeap... an Extension method is in order here

File: RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs [hunk 8]
```diff
@@ -94,13 +94,24 @@ public bool ValidateSelection(QualifiedSelection qualifiedSelection)
...
+                    return false;
+
+                // We've proved that there are no invalid statements contained in the selection. However, we need to analyze
                 // the statements to ensure they are not partial selections.
 
                 // The visitor will not return the results in a sorted manner, so we need to arrange the contexts in the same order.
-                var sorted = results.OrderBy(context => context.Start.StartIndex);
+                var blockStmtContexts = results as IList<VBAParser.BlockStmtContext> ?? results.ToList();
```
Comment (Vogel612) [inline RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs]: why require materialization here?

Comment (bclothier) [reply RetailCoder.VBE/Refactorings/ExtractMethod/ExtractMethodSelectionValidation.cs]: To avoid multiple enumeration (it's read twice subsequently)

File: RetailCoder.VBE/UI/RubberduckUI.resx [hunk 9]
```diff
@@ -2167,4 +2167,13 @@ Would you like to import them to Rubberduck?</value>
...
   </data>
+  <data name="ExtractMethod_CodePreviewCaption" xml:space="preserve">
+    <value>Code Preview extracted from {0}</value>
+  </data>
+  <data name="ExtractMethod_DefaultNewMethodName" xml:space="preserve">
+    <value>NewMethod</value>
+  </data>
+  <data name="ExtractMethod_NoneSelected" xml:space="preserve">
+    <value>(None)</value>
+  </data>
```
Comment (bclothier) [inline RetailCoder.VBE/UI/RubberduckUI.resx]: This duplicate file should not be committed and needs to be cleaned up in the repository.
